apps/arweave/src/ar_data_sync_worker_master.erl

```diff
@@ -254,99 +304,84 @@ schedule_read_range(Args, State) ->
 
-	PeerTasks2 = PeerTasks#peer_tasks{
-		active_count = ActiveCount,
-		max_active = MaxActive,
-		task_queue = TaskQueue
-	},
-	set_peer_tasks(Peer, PeerTasks2, State).
 
 %%--------------------------------------------------------------------
 %% Helpers
 %%--------------------------------------------------------------------
-update_counters(TaskState, Task, FormattedPeer, N) ->
-	Key = ets_key(TaskState),
-	ets:update_counter(?MODULE, Key, {2, N}, {Key,0}),
-	prometheus_gauge:inc(sync_tasks, [TaskState, Task, FormattedPeer], N).
-
-reset_counters(TaskState) ->
-	ets:insert(?MODULE, {ets_key(TaskState), 0}).
-
-get_task_count(TaskState) ->
-	case ets:lookup(?MODULE, ets_key(TaskState)) of
-		[] ->
-			0;
-		[{_, N}] ->
-			N
-	end.
-
-ets_key(TaskState) ->
-	list_to_atom(atom_to_list(TaskState) ++ "_tasks").
-
-get_ema(Peer) ->
-	case ets:lookup(?MODULE, {ema, Peer}) of
-		[] ->
-			undefined;
-		[{_, EMA}] ->
-			EMA
-	end.
-
-set_ema(Peer, EMA) ->
-	ets:insert(?MODULE, {{ema, Peer}, EMA}).
+update_queued_task_count(Task, FormattedPeer, N, State) ->
+	prometheus_gauge:inc(sync_tasks, [queued, Task, FormattedPeer], N),
+	State#state{ queued_task_count = State#state.queued_task_count + N }.
+update_scheduled_task_count(Worker, Task, FormattedPeer, N, State) ->
+	prometheus_gauge:inc(sync_tasks, [scheduled, Task, FormattedPeer], N),
+	Load = maps:get(Worker, State#state.worker_loads, 0) + N,
+	State2 = State#state{
+		scheduled_task_count = State#state.scheduled_task_count + N,
+		worker_loads = maps:put(Worker, Load, State#state.worker_loads)
+	},
+	State2.
 
-update_ema(Peer, IsOK, Milliseconds) ->
-	OldEMA = get_ema(Peer),
-	NewEMA = case {IsOK, OldEMA} of
-		{true, undefined} ->
-			Milliseconds / 1.0; %% convert to float
-		{true, OldEMA} ->
-			?EMA_ALPHA * Milliseconds + (1 - ?EMA_ALPHA) * OldEMA;
-		_ ->
-			OldEMA
-	end,
-	set_ema(Peer, NewEMA),
-	NewEMA.
+calculate_ema(OldEMA, false, _Value, _Alpha) ->
+	OldEMA;
+calculate_ema(OldEMA, true, Value, Alpha) ->
+	Alpha * Value + (1 - Alpha) * OldEMA.
 
 get_peer_tasks(Peer, State) ->
-	maps:get(Peer, State#state.peer_tasks, #peer_tasks{}).
-
-set_peer_tasks(Peer, PeerTasks, State) ->
-	State#state{ peer_tasks = maps:put(Peer, PeerTasks, State#state.peer_tasks) }.
-
-get_worker(#state{ all_workers = AllWorkerQ } = State) ->
-	{Worker, AllWorkerQ2} = cycle_worker(AllWorkerQ),
-	{Worker, State#state{all_workers = AllWorkerQ2}}.
-
-cycle_worker(WorkerQ) ->
-	{{value, Worker}, WorkerQ2} = queue:out(WorkerQ),
-	{Worker, queue:in(Worker, WorkerQ2)}.
+	maps:get(Peer, State#state.peer_tasks, #peer_tasks{peer = Peer}).
+
+set_peer_tasks(PeerTasks, State) ->
+	State#state{ peer_tasks =
+		maps:put(PeerTasks#peer_tasks.peer, PeerTasks, State#state.peer_tasks)
+	}.
+
+get_worker(State) ->
+	AverageLoad = State#state.scheduled_task_count / State#state.worker_count,
+	cycle_workers(AverageLoad, State).
+
+cycle_workers(AverageLoad, #state{ workers = Workers, worker_loads = WorkerLoads} = State) ->
```

Comment:
Better load balancing when sending work to `ar_data_sync_workers`. Keep cycling through the queue until we find a worker with less than average load.